README.md
# PlugAuth::Lite [![Build Status](https://secure.travis-ci.org/plicease/PlugAuth-Lite.png)](http://travis-ci.org/plicease/PlugAuth-Lite)

Pluggable (lite) authentication and authorization server.

# SYNOPSIS

command line:

    % plugauthlite

Mojolicious Plugin:

    use Mojolicious::Lite;
    
    plugin 'plug_auth_lite', 
      auth => sub {
        my($user, $pass) = @_;
        if($user eq 'optimus' && $pass eq 'matrix')
        { return 1; }
        else
        { return 0; }
      },
      authz => sub {
        my($user, $action, $resource) = @_;
        if($user eq 'optimus && $action eq 'open' && $resource =~ m{^/matrix})
        { return 1 }
        else
        { return 0 }
      };

Mojolicious App:

    use PlugAuth::Lite;
    my $app = PlugAuth::Lite->new({
      auth => sub {
        my($user, $pass) = @_;
        if($user eq 'optimus' && $pass eq 'matrix')
        { return 1; }
        else
        { return 0; }
      },
      authz => sub {
        my($user, $action, $resource) = @_;
        if($user eq 'optimus && $action eq 'open' && $resource =~ m{^/matrix})
        { return 1 }
        else
        { return 0 }
      },
    });

# DESCRIPTION

This distribution provides 

- [plugauthlite](http://search.cpan.org/perldoc?plugauthlite)

    [PlugAuth](http://search.cpan.org/perldoc?PlugAuth) compatible server in the form of a [Mojolicious::Lite](http://search.cpan.org/perldoc?Mojolicious::Lite) application.

- [Mojolicious::Plugin::PlugAuthLite](http://search.cpan.org/perldoc?Mojolicious::Plugin::PlugAuthLite)

    [Mojolicious](http://search.cpan.org/perldoc?Mojolicious) plugin that adds [PlugAuth](http://search.cpan.org/perldoc?PlugAuth) compatible routes to a new or 
    existing [Mojolicious](http://search.cpan.org/perldoc?Mojolicious) application.

- [PlugAuth::Lite](http://search.cpan.org/perldoc?PlugAuth::Lite)

    [Mojolicious](http://search.cpan.org/perldoc?Mojolicious) application with [PlugAuth](http://search.cpan.org/perldoc?PlugAuth) compatible routes that can be spawned
    from within a perl application.

In the future it will also contain a testing interface for testing authentication
and authorization rules in [Clustericious](http://search.cpan.org/perldoc?Clustericious) applications.

It has fewer prerequisites that the full fledged [PlugAuth](http://search.cpan.org/perldoc?PlugAuth) server (simply
[Mojolicious](http://search.cpan.org/perldoc?Mojolicious) and perl itself) but also fewer features (it notably lacks
the management interface).

# ATTRIBUTES

## auth

Subroutine reference to call to check authentication.  Passes in `($user, $pass)` should
return true for authenticated, false otherwise.

If not provided, all authentications fail.

## authz

Subroutine reference to call to check authorization.  Passes in `($user, $action, $resource)`
and should return true for authorized, false otherwise.

If not provided, all authorizations succeed.

## host

Subroutine reference to call to check host information.

# SEE ALSO

[plugauthlite](http://search.cpan.org/perldoc?plugauthlite),
[Mojolicious::Plugin::PlugAuthLite](http://search.cpan.org/perldoc?Mojolicious::Plugin::PlugAuthLite),
[Test::PlugAuth](http://search.cpan.org/perldoc?Test::PlugAuth),
[Clustericious](http://search.cpan.org/perldoc?Clustericious)

# AUTHOR

Graham Ollis <plicease@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.07      2013-10-26 08:39:39 -0400
  - added test for Test::PlugAuth
  - Mojo 4.50 deprecation compatability

0.06      2013-10-26 08:05:30 -0400
  - Mojolicious 4.50 compatability

0.05      2013-06-14 12:31:28 EST5EDT
  - Create an interface in Test::PlugAuth

0.04      2013-06-13 16:53:50 America/New_York
  - documentation

0.03      2013-06-13 16:25:05 America/New_York
  - undocumented host routes (experimental)
  - if you die in a callback server catches it and returns 503

0.02      2013-06-13 14:41:57 America/New_York
  - PlugAuth::Lite is now a Mojo app that uses PlugAuthLite plugin

0.01      2013-05-09 16:47:47 EST5EDT
  - initial version


lib/Test/PlugAuth.pm
package Test::PlugAuth;

use strict;
use warnings;
use v5.10;
use PlugAuth::Lite;
use Mojo::UserAgent;

# ABSTRACT: minimum PlugAuth server to test Clustericious apps against
# VERSION

=head1 SYNOPSIS

assuming you have a Clustericious app MyApp with authentication/authorization
directives that you need to test:

 use Test::Clustericious::Config;
 use Test::Clustericious;
 use Test::PlugAuth;
 
 my $auth = Test::PlugAuth->new(auth => {
   my($user,$pass) = @_;
   return $user eq 'gooduser' && $pass eq 'goodpass';
 });
 
 create_config_ok 'MyApp', { plug_auth => { url => $auth->url } };
 
 $t = Test::Clustericious->new('MyApp');
 $auth->apply_to_client_app($t->app);
 
 my $port = $t->ua->app_url->port;
 
 $t->get_ok("http://baduser:badpass\@localhost:$port/private")
   ->status_is(401);
 $t->get_ok("http://gooduser:goodpass\@localhost:$port/private")
   ->status_is(200);

=head1 DESCRIPTION

Provides a way to test a Clustericious application with a fake PlugAuth server
with reduced boilerplate

=head1 CONSTRUCTOR

=head2 Test::PlugAuth->new( $config )

Creates a new instance of Test::PlugAuth.  The $config is passed
directly into L<PlugAuth::Lite>.  See L<Mojolicious::Plugin::PlugAuthLite>
for details.

=cut

sub new
{
  my $class = shift;
  my $config = ref $_[0] ? $_[0] : {@_};
  my $self = bless {}, $class;
  
  $self->{app} = PlugAuth::Lite->new($config);
  $self->{ua}  = Mojo::UserAgent->new;
  $self->ua->app($self->app);
  
  $self->{url} = $self->ua->app_url->to_string;
  $self->{url} =~ s{/$}{};
  
  return $self;
}

=head1 ATTRIBUTES

=head2 ua

The L<Mojo::UserAgent> used to connect to the PlugAuth (lite) server.

=cut

sub ua  { shift->{ua}  }

=head2 app

The L<PlugAuth::Lite> instance of the PlugAuth server.

=cut

sub app { shift->{app} }

=head2 url

The (fake) url used to connect to the PlugAuth server with.  You MUST
connect to through the L<Mojo::UserAgent> above.

=cut

sub url { shift->{url} }

=head1 METHODS

=head2 $test_auth->apply_to_client_app( $client_app )

Given a Clustericious application C<$client_app>, this method will 
rewire our L<Mojo::UserAgent> for authentication requests to PlugAuth.

=cut

sub apply_to_client_app 
{
  my($self, $client_app) = @_;
  $client_app->helper(auth_ua => sub { $self->ua });
  return;
}

1;


lib/PlugAuth/Lite.pm
package PlugAuth::Lite;

use strict;
use warnings;
use v5.10;
use Mojo::Base qw( Mojolicious );

# ABSTRACT: Pluggable (lite) authentication and authorization server.
# VERSION

has 'auth';
has 'authz';
has 'host';

sub startup
{
  my($self, $config) = @_;

  $self->plugin('plug_auth_lite',
    auth  => $self->auth  // sub { 0 },
    authz => $self->authz // sub { 1 },
    host  => $self->host  // sub { 0 },
  );
}

=head1 SYNOPSIS

command line:

 % plugauthlite

Mojolicious Plugin:

 use Mojolicious::Lite;
 
 plugin 'plug_auth_lite', 
   auth => sub {
     my($user, $pass) = @_;
     if($user eq 'optimus' && $pass eq 'matrix')
     { return 1; }
     else
     { return 0; }
   },
   authz => sub {
     my($user, $action, $resource) = @_;
     if($user eq 'optimus && $action eq 'open' && $resource =~ m{^/matrix})
     { return 1 }
     else
     { return 0 }
   };

Mojolicious App:

 use PlugAuth::Lite;
 my $app = PlugAuth::Lite->new({
   auth => sub {
     my($user, $pass) = @_;
     if($user eq 'optimus' && $pass eq 'matrix')
     { return 1; }
     else
     { return 0; }
   },
   authz => sub {
     my($user, $action, $resource) = @_;
     if($user eq 'optimus && $action eq 'open' && $resource =~ m{^/matrix})
     { return 1 }
     else
     { return 0 }
   },
 });

=head1 DESCRIPTION

This distribution provides 

=over 4

=item L<plugauthlite>

L<PlugAuth> compatible server in the form of a L<Mojolicious::Lite> application.

=item L<Mojolicious::Plugin::PlugAuthLite>

L<Mojolicious> plugin that adds L<PlugAuth> compatible routes to a new or 
existing L<Mojolicious> application.

=item L<PlugAuth::Lite>

L<Mojolicious> application with L<PlugAuth> compatible routes that can be spawned
from within a perl application.

=back

In the future it will also contain a testing interface for testing authentication
and authorization rules in L<Clustericious> applications.

It has fewer prerequisites that the full fledged L<PlugAuth> server (simply
L<Mojolicious> and perl itself) but also fewer features (it notably lacks
the management interface).

=head1 SEE ALSO

L<plugauthlite>,
L<Mojolicious::Plugin::PlugAuthLite>,
L<Test::PlugAuth>,
L<Clustericious>

=cut

1;


lib/Mojolicious/Plugin/PlugAuthLite.pm
package Mojolicious::Plugin::PlugAuthLite;

use Mojo::Base qw( Mojolicious::Plugin );
use Mojo::ByteStream qw( b );
use v5.10;

# ABSTRACT: Add a minimal PlugAuth server to your Mojolicious application.
# VERSION

=head1 SYNOPSIS

 use Mojolicious::Lite
 
 plugin 'plug_auth_lite', 
   auth => sub {
     my($user, $pass) = @_;
     if($user eq 'optimus' && $pass eq 'matrix')
     { return 1; }
     else
     { return 0; }
   },
   authz => sub {
     my($user, $action, $resource) = @_;
     if($user eq 'optimus && $action eq 'open' && $resource =~ m{^/matrix})
     { return 1 }
     else
     { return 0 }
   };

=head1 DESCRIPTION

This plugin provides a very minimal but customizable L<PlugAuth> server which can
be included with your L<Mojolicious> application for L<Clustericious> applications
to authenticate against.  If you do not need specialized plugins for LDAP or DBI,
and if you do not need the user/group/resource management provided by a the full
featured L<PlugAuth> server then this plugin may be for you.

The script L<plugauthlite> included with this distribution provides PlugAuth
style authentication (but not authorization) using a simple Apache style password
file.

=head1 CONFIGURATION

=head2 auth

Subroutine which checks the authentication of a user.  It is passed two arguments,
the username and the password.  If they are authentic this call back should return
1.  Otherwise it should return 0.

=head2 authz

Subroutine which checks the authorization of a user.  It is passwd three arguments,
the username, action (usually a verb) and resource (usually the path part of a URL).
If the user is authorized for the action on that resource the call back should return
1.  Otherwise it should return 0.

=head2 url

The prefix to prepend to the standard PlugAuth API routes.  Usually the authentication
route is /auth and the authorization route is /authz, but if the PlugAuth.conf client
configuration is set to http://example.com/foo the client expects the authentication
route to be /foo/auth and the authorization route to be /foo/authz.  In this case you
would set this configuration item to '/foo'.

=head2 realm

The realm to use for HTTP Basic authentication.  The default is PlugAuthLite.

=head1 ROUTES

=head2 GET /auth

=over 4

=item * if username and password provided using BASIC authentication and are correct

Return 200 ok

=item * if username and password provided using BASIC authentication but are not correct

Return 403 not ok

=item * if username and password are not provided using BASIC authentication

Return 401 please authenticate

=back

=head2 GET /authz/user/#user/#action/(*resource)

=over 4

=item * if the given user (#user) is permitted to perform the given action (#action) on the given resource (*resource)

Return 200 ok

=item * otherwise

return 403 not ok

=back

=head1 METHODS

=head2 register

This method adds the routes to your application required to implement the PlugAuth
API.

=head1 LIMITATIONS

This implementation of the PlugAuth protocol does not support these features provided
by the full fledged L<PlugAuth> server:

=over 4

=item *

Groups

=item *

Management API for creating/removing/modifying users/groups/resources

=item *

Standard Clustericious routes like "/version" and "/status"

=item *

Clustericious configuration file (~/etc/PlugAuth.conf)

=item *

Support for L<PlugAuth> plugins (L<PlugAuth::Plugin>).

=item *

Probably many others.

=back

=head1 SEE ALSO

L<plugauthlite>,
L<PlugAuth::Lite>,
L<PlugAuth>

=cut

sub register
{
  my($self, $app, $conf) = @_;
  
  my $cb_auth  = $conf->{auth}  // sub { 0 };
  my $cb_authz = $conf->{authz} // sub { 1 };
  my $cb_host  = $conf->{host}  // sub { 0 };
  my $realm = $conf->{realm} // 'PlugAuthLite';
  my $base_url = $conf->{url} // $conf->{uri} // '';

  $app->routes->get("$base_url/auth" => sub {
    my $self = shift;
    eval {
      my $auth_header = $self->req->headers->authorization;
      unless($auth_header)
      {
        $self->res->headers->www_authenticate('Basic "$realm"');
        $self->render(text => 'please authenticate', status => 401);
        return;
      }
      my ($method,$str) = split / /,$auth_header;
      my ($user,$pw) = split /:/, b($str)->b64_decode;
      if($cb_auth->($user, $pw))
      {
        $self->render(text => 'ok', status => 200);
      }
      else
      {
        $self->render(text => 'not ok', status => 403);
      }
    };
    $self->render(text => 'not ok', status => 503) if $@;
  })->name('plugauth_auth');
  
  $app->routes->get("$base_url/authz/user/#user/#action/(*resource)" => { resource => '/' } => sub {
    my $self = shift;
    eval {
      my($user, $resource, $action) = map { $self->stash($_) } qw( user resource action );
      $resource =~ s{^/?}{/};
      if($cb_authz->($user, $action, $resource))
      {
        $self->render(text => 'ok', status => 200);
      }
      else
      {
        $self->render(text => 'not ok', status => 403);
      }
    };
    $self->render(text => 'not ok', status => 503) if $@;
  })->name('plugauth_authz');
  
  $app->routes->get("$base_url/host/#host/:tag" => sub {
    my $self = shift;
    eval {
      my ($host,$tag) = map $self->stash($_), qw/host tag/;
      if ($cb_host->($host,$tag)) {
        return $self->render(text => 'ok', status => 200);
      }
      return $self->render(text => 'not ok', status => 403);
    };
    $self->render(text => 'not ok', status => 503) if $@;
  })->name('plugauth_host');
  
  return;
}

1;
